drivers/video/video_stm32_venc.c

```diff
@@ -0,0 +1,899 @@
+
+	if (*vbuf == NULL) {
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+ISR_DIRECT_DECLARE(stm32_venc_isr)
+{
+	struct stm32_venc_ewl *inst = &ewl_instance;
+	const struct stm32_venc_config *config = inst->config;
+	uint32_t hw_handshake_status =
+		IS_BIT_SET(sys_read32(config->reg + BASE_HEncInstantInput), LOW_LATENCY_HW_ITF_EN);
+	uint32_t irq_status = sys_read32(config->reg + BASE_HEncIRQ);
+
+	inst->irq_status = irq_status;
+	inst->irq_cnt++;
+
+	if (hw_handshake_status == 0 && (irq_status & ASIC_STATUS_FUSE) != 0) {
+		sys_write32(ASIC_STATUS_FUSE | ASIC_IRQ_LINE, config->reg + BASE_HEncIRQ);
+		/* read back the IRQ status to update its value */
+		irq_status = sys_read32(config->reg + BASE_HEncIRQ);
+	}
+
+	if (irq_status != 0U) {
+		/* status flag is raised,
+		 * clear the ones that the IRQ needs to clear
+		 * and signal to EWLWaitHwRdy
+		 */
+		sys_write32(ASIC_STATUS_SLICE_READY | ASIC_IRQ_LINE, config->reg + BASE_HEncIRQ);
+	}
+
+	k_sem_give(&inst->complete);
+
+	return 0;
+}
+
+#define VENC_FORMAT_CAP(pixfmt)                \
+	{                                      \
+		.pixelformat = pixfmt,         \
+		.width_min = 48,               \
+		.width_max = 1920,             \
+		.height_min = 48,              \
+		.height_max = 1088,            \
+		.width_step = 16,              \
+		.height_step = 16,             \
+	}
+
+static const struct video_format_cap in_fmts[] = {
+	VENC_FORMAT_CAP(VIDEO_PIX_FMT_NV12),
+	VENC_FORMAT_CAP(VIDEO_PIX_FMT_RGB565),
+	{0},
+};
+
+static const struct video_format_cap out_fmts[] = {
+	VENC_FORMAT_CAP(VIDEO_PIX_FMT_H264),
+	{0},
+};
+
+static int stm32_venc_get_caps(const struct device *dev, struct video_caps *caps)
+{
+	if (caps->type == VIDEO_BUF_TYPE_INPUT) {
+		caps->format_caps = in_fmts;
+	} else {
+		caps->format_caps = out_fmts;
+	}
+
+	/* VENC produces full frames */
+	caps->min_line_count = caps->max_line_count = LINE_COUNT_HEIGHT;
+	caps->min_vbuf_count = 1;
+
+	return 0;
+}
+
+static DEVICE_API(video, stm32_venc_driver_api) = {
+	.set_format = stm32_venc_set_fmt,
+	.get_format = stm32_venc_get_fmt,
+	.set_stream = stm32_venc_set_stream,
+	.enqueue = stm32_venc_enqueue,
+	.dequeue = stm32_venc_dequeue,
+	.get_caps = stm32_venc_get_caps,
+};
+
+static void stm32_venc_irq_config_func(const struct device *dev)
+{
+	IRQ_DIRECT_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), stm32_venc_isr, 0);
+	irq_enable(DT_INST_IRQN(0));
+}
+
+static struct stm32_venc_data stm32_venc_data_0 = {};
+
+static const struct stm32_venc_config stm32_venc_config_0 = {
+	.reg = DT_INST_REG_ADDR(0),
+	.pclken = {.bus = DT_INST_CLOCKS_CELL(0, bus), .enr = DT_INST_CLOCKS_CELL(0, bits)},
+	.reset = RESET_DT_SPEC_INST_GET_BY_IDX(0, 0),
+	.irq_config = stm32_venc_irq_config_func,
+};
+
+static void RISAF_Config(void)
+{
+	/* Define and initialize the master configuration structure */
+	RIMC_MasterConfig_t RIMC_master = {0};
+
+	/* Enable the clock for the RIFSC (RIF Security Controller) */
+	__HAL_RCC_RIFSC_CLK_ENABLE();
```

Comment:
Sorry I do not know the STM32N6 hardware very well. Out of curiosity, is this unrelated to this?
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/reset/st%2Cstm32-rcc-rctl.yaml

Reply:
Understood, good to learn about the connection.